openwisp_monitoring/device/admin.py

```diff
@@ -139,6 +154,12 @@ def get_form(self, request, obj=None, **kwargs):
         """
         Adds the help_text of DeviceMonitoring.status field
         """
+        health_field = 'metric_health' in self.fields
+        if obj and obj.monitoring.status != 'problem' and health_field:
+            self.fields.remove('metric_health')
+        elif obj and obj.monitoring.status == 'problem' and health_field is False:
+            self.fields.insert(self.fields.index('last_ip'), 'metric_health')
+            self.readonly_fields.append('metric_health')
```

Comment:
the approach is not correct.

You have to understand that both `self.fields` and `self.readonly_fields` are class attributes which point to lists, which are mutable data structures. When you modify at runtime a list which is a class attribute (hence a global value), you are modifying it for the whole application.

This approach can lead to inconsistent behavior.

You should instead override these two methods:

- https://docs.djangoproject.com/en/3.1/ref/contrib/admin/#django.contrib.admin.ModelAdmin.get_fields
- https://docs.djangoproject.com/en/3.1/ref/contrib/admin/#django.contrib.admin.ModelAdmin.get_readonly_fields